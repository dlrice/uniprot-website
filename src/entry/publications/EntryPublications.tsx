import React, { FC } from 'react';
import { uniq } from 'lodash';
import { Loader, Publication, DataList } from 'franklin-sites';
import { LiteratureForProteinAPI } from '../../literature/types/LiteratureTypes';

const EntryPublications: FC<{
  accession: string;
  data: LiteratureForProteinAPI[] | null;
  total: number;
  handleLoadMoreItems: () => void;
}> = ({ accession, data, total, handleLoadMoreItems }) => {
  if (!data || data.length === 0) {
    return <Loader />;
  }

  return (
    <section>
      <div style={{ height: '80vh' }}>
        <h2>Publications for {accession}</h2>
        {/* The height css will be removed after Franklin DataList is updated */}
        <DataList
          idKey="id"
          data={data}
          dataRenderer={({
            literatureMappedReference,
            publicationSource,
            uniProtReference,
            categories,
            literatureEntry,
          }: LiteratureForProteinAPI) => {
            let infoListData;
            if (literatureMappedReference) {
              const {
                annotation,
                source,
                sourceId,
              } = literatureMappedReference;
              infoListData = [
                {
                  title: 'Annotation',
                  content: annotation,
                },
                {
                  title: 'Mapping Source',
                  content: `${source}:${sourceId}`,
                },
                {
                  title: 'Source',
                  content: publicationSource,
                },
              ];
            } else if (uniProtReference) {
              const {
                referencePositions,
                referenceComments,
              } = uniProtReference;
              infoListData = [
                {
                  title: 'Cited for',
                  content: referencePositions,
                },
                {
                  title: 'Tissue',
                  content: referenceComments && (
                    <ul className="no-bullet">
                      {referenceComments.map(comment => (
                        <li key={comment.value}>{comment.value}</li>
                      ))}
                    </ul>
                  ),
                },
                {
                  title: 'Categories',
                  content: categories && (
                    <ul className="no-bullet">
                      {uniq(categories).map(category => (
                        <li key={category}>{category}</li>
                      ))}
                    </ul>
                  ),
                },
                {
                  title: 'Source',
                  content: publicationSource,
                },
              ];
            }
            if (literatureEntry) {
              const { literatureAbstract, statistics } = literatureEntry;
              return (
                <Publication
                  {...literatureEntry}
                  abstract={literatureAbstract}
                  infoData={infoListData}
                  statistics={statistics && statistics}
                />
              );
            }
            return (
              uniProtReference && (
                <Publication
                  {...uniProtReference.citation}
                  infoData={infoListData}
                />
              )
            );
          }}
          onLoadMoreItems={handleLoadMoreItems}
          hasMoreData={total > data.length}
        />
      </div>
    </section>
  );
};

export default EntryPublications;
