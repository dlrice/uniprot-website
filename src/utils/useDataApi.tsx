/* eslint-disable @typescript-eslint/no-explicit-any */
import { useState, useEffect } from 'react';
import fetchData from './fetchData';

const useDataApi = (url: string, headers: any) => {
  const [data, setData] = useState<any>({});

  useEffect(() => {
    const getData = async () => {
      const result = await fetchData(url, headers);
      setData(result.data);
    };
    getData();
  }, [url]);

  return data;
};

export default useDataApi;
