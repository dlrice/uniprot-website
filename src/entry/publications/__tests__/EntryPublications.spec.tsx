import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import EntryPublications from '../EntryPublications';
import mockPublicationsData from '../__mocks__/entryPublicationsData.json';

import useDataApi from '../../../hooks/useDataApi';

jest.mock('../../../hooks/useDataApi', () => ({
  __esModule: true, // this makes it work
  default: jest.fn(() => ({
    loading: false,
    data: mockPublicationsData,
    headers: {
      'x-totalrecords': mockPublicationsData.results.length,
    },
  })),
}));

describe('EntryPublications tests', () => {
  it('should render', () => {
    // It seems the setState in useEffect causes this to hang
    // render(
    //   <Router>
    //     <EntryPublications accession="P05067" selectedFacets={[]} />
    //   </Router>
    // );
    // expect(useDataApi).toHaveBeenCalled();
  });
});
