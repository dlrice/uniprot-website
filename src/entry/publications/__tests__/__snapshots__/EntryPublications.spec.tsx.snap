// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryPublications tests should render 1`] = `
<DocumentFragment>
  <section>
    <div
      style="height: 80vh;"
    >
      <h2>
        Publications for P05067
      </h2>
      <div
        class="data-loader__wrapper"
      >
        <div
          class="data-loader__scroll-container"
        >
          <div
            class="card "
          >
            <div
              class="card__content"
            >
              <section
                class="publication"
              >
                <h4>
                  The precursor of Alzheimer's disease amyloid A4 protein resembles a cell- surface receptor.
                </h4>
                <section
                  class="publication__authors"
                >
                  <a
                    href="/"
                  >
                    Kang J.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Lemaire H.-G.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Unterbeck A.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Salbaum J.M.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Masters C.L.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Grzeschik K.-H.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Multhaup G.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Beyreuther K.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Mueller-Hill B.
                  </a>
                </section>
                <section
                  class="publication__columns"
                >
                  <section
                    class="publication__columns__main"
                  >
                    <section>
                      <button
                        class="button tertiary"
                        type="button"
                      >
                        Show abstract
                      </button>
                    </section>
                    <div
                      class="info-list"
                    >
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Cited for
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          NUCLEOTIDE SEQUENCE [MRNA] (ISOFORM APP695)
                        </div>
                      </div>
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Tissue
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          <ul
                            class="no-bullet"
                          >
                            <li>
                              Brain
                            </li>
                          </ul>
                        </div>
                      </div>
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Categories
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          <ul
                            class="no-bullet"
                          >
                            <li>
                              Sequence
                            </li>
                          </ul>
                        </div>
                      </div>
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Source
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          Swiss-Prot
                        </div>
                      </div>
                    </div>
                  </section>
                  <section
                    class="publication__columns__side"
                  >
                    <ul
                      class="no-bullet"
                    >
                      <li>
                        <a
                          href="//www.ncbi.nlm.nih.gov/pubmed/2881207"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          PubMed
                        </a>
                      </li>
                      <li>
                        <a
                          href="//europepmc.org/article/MED/2881207"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          Europe PMC
                        </a>
                      </li>
                      <li>
                        <section
                          class="publication__statistics"
                        >
                          <div>
                            Cited in other entries:
                          </div>
                          <span
                            class="bubble--medium colour-sea-blue"
                          >
                            1
                          </span>
                        </section>
                      </li>
                    </ul>
                  </section>
                </section>
              </section>
            </div>
          </div>
          <div
            class="card "
          >
            <div
              class="card__content"
            >
              <section
                class="publication"
              >
                <h4>
                  Constitutive and regulated alpha-secretase cleavage of Alzheimer's amyloid precursor protein by a disintegrin metalloprotease.
                </h4>
                <section
                  class="publication__authors"
                >
                  <a
                    href="/"
                  >
                    Lammich S.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Kojro E.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Postina R.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Gilbert S.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Pfeiffer R.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Jasionowski M.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Haass C.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Fahrenholz F.
                  </a>
                </section>
                <section
                  class="publication__columns"
                >
                  <section
                    class="publication__columns__main"
                  >
                    <section>
                      <button
                        class="button tertiary"
                        type="button"
                      >
                        Show abstract
                      </button>
                    </section>
                    <div
                      class="info-list"
                    >
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Annotation
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          Pathway
                        </div>
                      </div>
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Mapping Source
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          Reactome:R-HSA-9010028
                        </div>
                      </div>
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Source
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          Computationally mapped
                        </div>
                      </div>
                    </div>
                  </section>
                  <section
                    class="publication__columns__side"
                  >
                    <ul
                      class="no-bullet"
                    >
                      <li>
                        <a
                          href="//www.ncbi.nlm.nih.gov/pubmed/10097139"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          PubMed
                        </a>
                      </li>
                      <li>
                        <a
                          href="//europepmc.org/article/MED/10097139"
                          rel="noopener noreferrer"
                          target="_blank"
                        >
                          Europe PMC
                        </a>
                      </li>
                      <li>
                        <section
                          class="publication__statistics"
                        >
                          <div>
                            Cited in other entries:
                          </div>
                          <span
                            class="bubble--medium colour-sea-blue"
                          >
                            1
                          </span>
                        </section>
                      </li>
                    </ul>
                  </section>
                </section>
              </section>
            </div>
          </div>
          <div
            class="card "
          >
            <div
              class="card__content"
            >
              <section
                class="publication"
              >
                <h4>
                  The phosphotyrosine interaction domains of X11 and FE65 bind to distinct sites on the YENPTY motif of amyloid precursor protein.
                </h4>
                <section
                  class="publication__authors"
                >
                  <a
                    href="/"
                  >
                    Borg J.-P.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Ooi J.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Levy E.
                  </a>
                  , 
                  <a
                    href="/"
                  >
                    Margolis B.
                  </a>
                </section>
                <section
                  class="publication__columns"
                >
                  <section
                    class="publication__columns__main"
                  >
                    <div
                      class="info-list"
                    >
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Cited for
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          INTERACTION WITH APBA1 AND APBB1MUTAGENESIS OF TYR-728; TYR-757; ASN-759 AND TYR-762
                        </div>
                      </div>
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Categories
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          <ul
                            class="no-bullet"
                          >
                            <li>
                              Interaction
                            </li>
                            <li>
                              Pathol
                            </li>
                          </ul>
                        </div>
                      </div>
                      <div
                        class="info-list__item"
                      >
                        <div
                          class="info-list__item__title"
                        >
                          <h5
                            class="bold"
                          >
                            Source
                          </h5>
                        </div>
                        <div
                          class="info-list__item__content"
                        >
                          Swiss-Prot
                        </div>
                      </div>
                    </div>
                  </section>
                  <section
                    class="publication__columns__side"
                  >
                    <ul
                      class="no-bullet"
                    />
                  </section>
                </section>
              </section>
            </div>
          </div>
          <div
            class="data-loader__loading"
          >
            Loading...
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
